src/analysis/subspaces/computeElementalMatrix.m

```diff
@@ -31,73 +45,278 @@
 
 metIDs = reshape(metIDs, length(metIDs),1);
 
-Ematrix = zeros(length(metIDs), 6);
-for n = 1:length(metIDs)
-    i = metIDs(n);
-    formula = model.metFormulas(i);
-    [compounds, tok] = regexp(formula, '([A-Z][a-z]*)(\d*)', 'match', 'tokens');
-    tok = tok{1,1};
-    for j = 1:length(tok) % go through each token.
-        t = tok{1,j};
-        comp = t{1,1};
-        q = str2num(t{1,2});
-        if (isempty(q))
-            q = 1;
+if ~genericFormula
+    elements = {'C', 'N', 'O', 'H', 'P', 'Other'};
+    Ematrix = zeros(length(metIDs), 6);
+    for n = 1:length(metIDs)
+        i = metIDs(n);
+        formula = model.metFormulas(i);
+        [compounds, tok] = regexp(formula, '([A-Z][a-z]*)(\d*)', 'match', 'tokens');
+        tok = tok{1,1};
+        for j = 1:length(tok) % go through each token.
+            t = tok{1,j};
+            comp = t{1,1};
+            q = str2num(t{1,2});
+            if (isempty(q))
+                q = 1;
+            end
+            switch comp

```

Comment:
Would it make sense to update this switch into something shorter (along with the later comment of having a general Elemental Weights Map loaded from a database?
e.g. something along the lines:
```Matlab
persistent ElementalWeightMatrix
if isempty(ElementalWeightMatrix)
elementalWeightMatrix = getElementalWeightMatrix(); %Read in the data from the weight matrix
end
elements = {'C', 'N', 'O', 'H', 'P', 'Other'};
 for n = 1:length(metIDs)
      i = metIDs(n);
      formula = model.metFormulas(i);
      [compounds, tok] = regexp(formula, '([A-Z][a-z]*)(\d*)', 'match', 'tokens');
      tok = tok{1,1};
        for j = 1:length(tok) % go through each token.
            comp = t{1,1};
            q = 1;
            q = str2num(t{1,2});
            elementpos = ismember(elements,comp);
            isspecialElement = any(elementpos(1:5));
            isElement = any(ismember(ElementalWeightMatrix(:,1),comp));
            if isspecialElement
                  EMatrix(elementpos) = Ematrix(elementpos) + q;
            elseif isElement
                  EMatrix(6) = Ematrix(6) + q;
            else
                    if warnings
                        display('Warning');
                        display(formula)
                        display(comp);
                    end
          end
     end
end
```

Reply:
Hi @tpfau, thanks for the thoughtful comments. When I did this, I just tried to minimize the changes to the original code and keep the original output behavior. It would be nice to have what you suggested. I think `parse_Atomic_Weights_and_Isotopic_Compositions_for_All_Elements` contains such information but needs some processing as in `getMolecularMass`.

Reply:
I actually did not see this function before. 
as for getMolecularMass:
This could also be very much reduced given that the isotopic distribution is (for all I know) also the distribution in organic molecules. So you can really just add up weighted sums of all isotopes with non NaN Isotopic compositions.
The only manual modification would be to modify the result from `parse...` to contain H for Hydrogen, Deuterium and Tritium... 
Do you want to go for this?

Reply:
I am not sure. I remember that the weights by `getMolecularMass` are a bit different from the standard molecular weights for some elements? But this should be the way to go.
And more generally, I think many of the functions here can be combined and re-organized to have a structure like:
- one function to find elemental compositions (with flags allowing only specific elements or generic elements or supporting brackets or not)
- one function to get the molecular weight of chemical elements (real number for true elements, NaN for hypothetical).
- one function to get molecular weight of metabolites (simply multiplying the output from the above two), with the part for hypothetical elements returned as residual formula not accounting for MWs as you suggested
- one function to check mass charge balance (simply multiplying the output from the first function by the S-matrix)

Reply:
I completely agree.
And yes, we would need to make sure, that the standard atomic weights are considered (that was my thought about obtaining the data from a database). 

A suggestion from my side:
This is all concerned with massBalance, so it should go into `reconstruction/modelGeneration/massBalance`
To keep backward compatability, existing functions should be modified to use computation from the "basic" functions you suggest.
We currently have the following functions:
- `checkMassChargeBalance` - this should (imo) be the overall Mass/Charge balance check function. This can be adapted for speedup and other improvements, but I would keep it as a basic function. 
- `getMolecularMass`: This function by default assumes, that only the predominant isotopic species is present (which is in contrast to what standard weights for molecules assume). In the code it is only ever called with this option, and a quick test shows that the alternative option doesn't work (at least it gives me an error). So I would suggest, to replace code in this function to actually use the standard weights (i.e. isotope mixture), yes this might change some things, but I would claim that this is to be preferred than obtaining potentially wrong results.
Also if there are undefined elements in the formula, without a special flag I would return NaN or [NaN, KnownWeight, UnknownElements], to make clear, that the MW is unclear and indicate the known weight parts.

- About the function for atomic element weights:
  I'd suggest to create a function that uses the output from `parse_...` to define the element weights. This function should have an option to take the average isotopoic weight or the weight of the predominant isotope (the former being the default).
One issue here: What do we do about H/D/T ? My suggestion would be: Use the average isotopic for H, and specific isotope weight for D and T (if averages are requested), assuming that D and T in a formula are quite clear to be D or T and not a general H.
- to get the elemental composition:
  there is currently no function for this, so your `parseGenericFormula` would be a good option  here. I'd actually suggest to rename it to `getElementalComposition`, and by default parse everything (including brackets), without allowing negative numbers. Only if a `formulaContainsCharge` flag (or something similar) is set, this function should assume a specific element name for `charge` and add it as an element. But otherwise I would really prohibit negative numbers. 

Does this sound reasonable?